### PR TITLE
[analytics-amplitude][ncl] Add API tests 

### PR DESCRIPTION
**Pull request by kkafar**
# Why

Amplitude module lacks test tools. 

# How

Added separate screen in API section of BareExpo app for testing Amplitude module. 

Note: testing requires Amplitude API key

# Test Plan

android + ios
Tested on demo Amplitude account 

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.io and README.md).
- [x] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).

### CODE REVIEW

**Comment by kkafar** [comment]
cc @lukmccall @bbarthec

File: apps/native-component-list/src/screens/Amplitude/AmplitudeScreen.tsx [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+        },
+        () => {
+          this.setState({ error: true, errorMessage: 'Failed to initialize module' });
+          this._deferredErrorCleanup();
+        }
+      );
+    } else {
+      this.setState({ error: true, errorMessage: 'Invalid api key' });
+      this._deferredErrorCleanup();
+    }
```
Comment (bbarthec) [inline apps/native-component-list/src/screens/Amplitude/AmplitudeScreen.tsx:66]:
```suggestion
    if (this.state.clientAPIKey) {
      try {
        await Amplitude.initializeAsync(this.state.clientAPIKey);
        this.setState({ result: 'Module initialized', initialized: true });
        this._deferredResultCleanup();
      } catch {
        this.setState({ error: 'Failed to initialize module' });
        this._deferredErrorCleanup();
      }
    } else {
      this.setState({ error: 'Invalid api key' });
      this._deferredErrorCleanup();
    }
```

File: apps/native-component-list/src/screens/Amplitude/AmplitudeScreen.tsx [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+        this._deferredResultCleanup();
+      },
+      () => {
+        this.setState({
+          error: true,
+          errorMessage: 'Failed to set UserId. Make sure that provided API key is valid.',
+        });
+        this._deferredErrorCleanup();
+      }
+    );
```
Comment (bbarthec) [inline apps/native-component-list/src/screens/Amplitude/AmplitudeScreen.tsx:82]:
```suggestion
    try {
      await Amplitude.setUserIdAsync(this.testUserId);
      this.setState({ result: `User ID set to ${this.testUserId}` });
      this._deferredResultCleanup();
    } catch {
      this.setState({ error: 'Failed to set UserId. Make sure that provided API key is valid.' });
      this._deferredErrorCleanup();
    }
```

File: apps/native-component-list/src/screens/Amplitude/AmplitudeScreen.tsx [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+        this._deferredResultCleanup();
+      },
+      () => {
+        this.setState({
+          error: true,
+          errorMessage: 'Failed to log event. Make sure that provided API key is valid.',
+        });
+        this._deferredErrorCleanup();
+      }
+    );
```
Comment (bbarthec) [inline apps/native-component-list/src/screens/Amplitude/AmplitudeScreen.tsx:117]:
```suggestion
    try {
      await Amplitude.logEventWithPropertiesAsync(this.testEventName, {
      this.setState({ result: `${this.testEventName} with mock properties logged` });
      this._deferredResultCleanup();
    } catch {
      this.setState({ error: 'Failed to log event. Make sure that provided API key is valid.' });
      this._deferredErrorCleanup();
    }
```

File: apps/native-component-list/src/screens/Amplitude/AmplitudeScreen.tsx [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+        this._deferredResultCleanup();
+      },
+      () => {
+        this.setState({
+          error: true,
+          errorMessage: 'Failed to add groups. Make sure that provided API key is valid.',
+        });
+        this._deferredErrorCleanup();
+      }
+    );
```
Comment (bbarthec) [inline apps/native-component-list/src/screens/Amplitude/AmplitudeScreen.tsx:133]:
```suggestion
    try {
      await Amplitude.setGroupAsync(this.testGroupType, ['groupName1', 'groupName2']);
      this.setState({ result: 'Groups added' });
      this._deferredResultCleanup();
    } catch {
      this.setState({ error: 'Failed to add groups. Make sure that provided API key is valid.' });
      this._deferredErrorCleanup();
    }
```

File: apps/native-component-list/src/screens/Amplitude/AmplitudeScreen.tsx [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+        <View style={styles.button}>
+          <Button title="Set user id" onPress={this._setUserIdAsync} />
+        </View>
+        <View style={styles.button}>
+          <Button title="Set groups for user" onPress={this._setGroupAsync} />
+        </View>
+        <TrackingOptionsSelector />
+        {this.state.error && (
+          <View style={[styles.textView, styles.errorView]}>
+            <Text style={styles.errorText}>{this.state.errorMessage}</Text>
```
Comment (bbarthec) [inline apps/native-component-list/src/screens/Amplitude/AmplitudeScreen.tsx:180]:
```suggestion
            <Text style={styles.errorText}>{this.state.error}</Text>
```

File: apps/native-component-list/src/screens/Amplitude/TrackingOptionsSelector.tsx [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+      disableLatLng: true,
+      disableOSName: true,
+      disableOSVersion: true,
+      disablePlatform: true,
+      disableRegion: true,
+      disableVersionName: true,
+    },
+    showList: false,
+    result: undefined,
+    error: undefined,
```
Comment (bbarthec) [inline apps/native-component-list/src/screens/Amplitude/TrackingOptionsSelector.tsx:55]:
```suggestion
```

File: apps/native-component-list/src/screens/Amplitude/TrackingOptionsSelector.tsx [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+    <View
+      style={{
+        flexDirection: 'row',
+        alignItems: 'center',
+        justifyContent: 'space-between',
+        paddingVertical: 5,
+        borderBottomWidth: 1.0 / PixelRatio.get(),
+        borderBottomColor: '#cccccc',
+      }}>
+      <Text style={{ flex: 1, fontSize: 16 }}>{title}</Text>
```
Comment (bbarthec) [inline apps/native-component-list/src/screens/Amplitude/TrackingOptionsSelector.tsx:99]: Extract style to `StyleSheet.create`
```suggestion
    <View
      style={styles.toggleConatainer}>
      <Text style={styles.toggleLabel}>{title}</Text>
```

File: apps/native-component-list/src/screens/Amplitude/TrackingOptionsSelector.tsx [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+          }))
+        }
+      />
+    </View>
+  );
+
+  _toggleListVisibility = () => {
+    this.setState(state => {
+      return { showList: !state.showList };
+    });
```
Comment (bbarthec) [inline apps/native-component-list/src/screens/Amplitude/TrackingOptionsSelector.tsx:115]:
```suggestion
    this.setState(state => ({ showList: !state.showList });
```

File: apps/native-component-list/src/screens/Amplitude/TrackingOptionsSelector.tsx [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+          this.setState({ result: undefined });
+        }, 5000);
+      },
+      () => {
+        this.setState({ error: 'Failed to send request for tracking options change' });
+        setTimeout(() => {
+          this.setState({ error: undefined });
+        }, 5000);
+      }
+    );
```
Comment (bbarthec) [inline apps/native-component-list/src/screens/Amplitude/TrackingOptionsSelector.tsx:137]:
```suggestion
    try {
      await setTrackingOptionsAsync(this.state.options);
      this.setState({ result: 'Request for tracking options change has been sent' });
      setTimeout(() => this.setState({ result: undefined }), 5000);
    } catch {
      this.setState({ error: 'Failed to send request for tracking options change' });
      setTimeout(() => this.setState({ error: undefined }), 5000);
    }
```

File: apps/native-component-list/src/screens/Amplitude/TrackingOptionsSelector.tsx [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+  };
+
+  render() {
+    return (
+      <View style={{ marginTop: 20 }}>
```
Comment (bbarthec) [inline apps/native-component-list/src/screens/Amplitude/TrackingOptionsSelector.tsx:142]:
```suggestion
      <View style={styles.container}>
```

File: apps/native-component-list/src/screens/Amplitude/TrackingOptionsSelector.tsx [hunk 11]
```diff
@@ -0,0 +1,263 @@
...
+        {this.state.showList &&
+          this._renderToggle({
+            title: 'Disable Region',
+            valueName: 'disableRegion',
+          })}
+        {this.state.showList &&
+          this._renderToggle({
+            title: 'Disable version name',
+            valueName: 'disableVersionName',
+          })}
```
Comment (bbarthec) [inline apps/native-component-list/src/screens/Amplitude/TrackingOptionsSelector.tsx:228]: Use [`Fragments`](https://reactjs.org/docs/fragments.html#short-syntax) and reduce boilerplate:
```suggestion
        {this.state.showList && (
          <>
            {
              [
            	{ title: 'Disable Adid', valueName: 'disableAdid' },
            	{ title: 'Disable Carrier', valueName: 'disableCarrier' },
            	{ title: 'Disable City', valueName: 'disableCity' },
            	{ title: 'Disable Country', valueName: 'disableCountry' },
            	{ title: 'Disable DMA', valueName: 'disableDMA' },
                { title: 'Disable Device brand', valueName: 'disableDeviceBrand' }
            	{ title: 'Disable Device manufacturer', valueName: 'disableDeviceManufacturer' },
            	{ title: 'Disable Device model', valueName: 'disableDeviceModel' },
            	{ title: 'Disable IDFV', valueName: 'disableIDFV' },
            	{ title: 'Disable IP Address', valueName: 'disableIPAddress' },
            	{ title: 'Disable Language', valueName: 'disableLanguage' },
            	{ title: 'Disable Geo?', valueName: 'disableLatLng' },
                { title: 'Disable OS Name', valueName: 'disableOSName' },
                { title: 'Disable OS version', valueName: 'disableOSVersion' },
                { title: 'Disable Platform', valueName: 'disablePlatform' },
                { title: 'Disable Region', valueName: 'disableRegion' },
                { title: 'Disable version name', valueName: 'disableVersionName' },
              ].map(this._renderToggle)
            }
          </>
        )}
```
You can extract these toggles definitions into some meaningful variable 😉

**Review by bbarthec — CHANGES_REQUESTED** [review]
Overall it looks interesting 👍 Good work on it 😉 
I've got a couple remarks apart from what's in the suggestions comments:
- keep up the good coding & ReactNative practices everywhere all the time
- if you repeat something for the 2nd time then it's the time to use some loop or refactor your approach 😉

**Review by bbarthec — APPROVED** [review]
Looks awesome 🎉

**Pull request merged**